### DISCUSSION TITLE
MVVM frameworks for Uno

### DISCUSSION THREAD

**Discussion by irenuncio** [unoplatform/uno#9806] [Q&A] [answered]
Hi everybody,

I'm just starting with Uno platform and would like to know if a MVVM framework should be used for non-trivial applications.

Is Uno compatible with widely used MVVM frameworks? In that case, which one would be a better fit for Uno, taking into account its design philosophy? Any recommendation (PRISM, MVVM Light, etc.)?

Kind regards,

Ignacio Renuncio.-

**Comment by seanocali** [comment, marked as answer]
The Microsoft MVVM Toolkit will work.

[MVVM](https://learn.microsoft.com/en-us/windows/communitytoolkit/mvvm/introduction)

**Reply by joyful-jisu** [reply]
Does Microsoft.CommunityToolkit.Mvvm compatible with IOS, WASM, Android?
I am currently using Uno and CommunityToolkit.Mvvm, Window platform works fine but other platform can't.

**Reply by hawkerm** [reply]
@joyful-jisu the MVVM Toolkit is .NET UX Platform and platform independent, it should work anywhere without issue (I've used it with Uno WASM without issue, for instance).

You're going to have to be a bit more specific with your setup, what's not working, and what you've tried to do already.

**Reply by dansiegel** [reply]
@joyful-jisu it is worth noting that the Extensions Template actually uses the CommunityToolkit out of the box when you disable Uno.Extensions.Reactive

**Reply by joyful-jisu** [reply]
> @joyful-jisu the MVVM Toolkit is .NET UX Platform and platform independent, it should work anywhere without issue (I've used it with Uno WASM without issue, for instance).
> 
> You're going to have to be a bit more specific with your setup, what's not working, and what you've tried to do already.

@hawkerm @dansiegel Thanks for reply. I was trying to bind RelayCommand on 'Command' attribute by using  Mvvm module on CommunityTookit. It worked on Windows platform but other platform was not. The visual studio log showed something like 'Unoplatform could not find Mvvm module'.
This is my first time developing a desktop application on Windows, so it was difficult to describe my situation. Sorry for asking vaguely.

**Reply by hawkerm** [reply]
@joyful-jisu sounds like you've only added the NuGet package to your UWP head and not the heads for the other platforms as well (as would be required).

**Reply by joyful-jisu** [reply]
@hawkerm 
You are right. I did some dumb mistake. I installed MVVM Toolkit on all platform.
It worked on Wasm, Windows, Android(I don't have other platform right now, so i only checked WASM and windows)
Thanks!

**Comment by ToddGlodek** [comment]
There is a REALLY big problem,  however, resulting from [EPIC] Controls initialization sequence ( #3519 ).   No matter wich MVVM framework you use, the problem you will have is that the DataContextChanged event and the Loaded event do not fire in the same sequence on WASM as they do on WinUI/UWP.   So, when your page/control is initializing - the ViewModel will always be null on WASM when DataContextChanged event is fired and often times also when Loaded event is called as well.

So  - you're going to find that on WinUI/UWP all your screens work fine and all the data bindings are OK.   Then on WASM you might end up with pages/controls that did not render any of the data at all.

@jeromelaban  made this excellent live-stream (https://www.youtube.com/watch?v=QkMXaBAcEd8) that proposed a fix.   I tried to adapted his solution as follows - but I haven't gotten it to work 100% of the time yet.    Essentially, it provies a delay - after which DataContextChanged & Loaded events get called a SECOND TIME (after the view model has been initialized).
```

public interface IUnoWasmIssue3519Hack
{

    public static readonly int DELAY = 500;

    void OnUnoWasmIssue3519_DataContextChanged(FrameworkElement sender, DataContextChangedEventArgs args);

    void OnUnoWasmIssue3519_Loaded(object sender, RoutedEventArgs e);

}

```
I Attached that interface to all your various controls - then I added the following in constructor of those classes 


```
        DispatcherQueue.TryEnqueue( () => {
                
                Task.Delay(IUnoWasmIssue3519Hack.DELAY).ContinueWith(rslt => {

                    this.Log().LogDebug( $"{this.GetType().Name} IUnoWasmIssue3519Hack *** Invoked");

#if HAS_UNO

                    DataContextChangedEventArgs eventArgs = new DataContextChangedEventArgs(DataContext);

                    try
                    {
                        this.OnUnoWasmIssue3519_DataContextChanged(this, eventArgs);

                        this.OnUnoWasmIssue3519_Loaded(this, new RoutedEventArgs());
                    }
                    catch (Exception ex) {

                        MyExceptionAbstract.HandleException(Logger, ex, "PanelAbstract IUnoWasmIssue3519Hack Invokation");
                    
                    }

#endif

                });            

        });
```
    
This is only a TEMPORARY HACK until #3519 gets fixed.    I hope you have better luck than I did.
